File: prisma/seed.mjs
import { PrismaClient } from '@prisma/client';
import * as https from 'https';

const prisma = new PrismaClient();

const getAccessToken = async () => {
  const clientId = '98b28229eb804c5f92211f35516a7e3c';
  const clientSecret = 'a0ecf390e2dd4ef5a73036ddbb5e9629';
  const payload = `grant_type=client_credentials&client_id=${clientId}&client_secret=${clientSecret}`;

  const options = {
    hostname: 'accounts.spotify.com',
    path: '/api/token',
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
  };

  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        try {
          const token = JSON.parse(data).access_token;
          resolve(token);
        } catch (parseError) {
          reject(parseError);
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.write(payload);
    req.end();
  });
};

const searchPlaylistId = async (token, playlistName) => {
  const params = new URLSearchParams({
    q: playlistName,
    type: 'playlist',
    limit: '1',
  });

  const options = {
    hostname: 'api.spotify.com',
    path: `/v1/search?${params.toString()}`,
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
    },
  };

  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        resolve(JSON.parse(data));
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.end();
  });
};

const searchPlaylist = async (token, playlistId) => {
  const options = {
    hostname: 'api.spotify.com',
    path: `/v1/playlists/${playlistId}`,
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
    },
  };

  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        resolve(JSON.parse(data));
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.end();
  });
};

const main = async () => {
  try {
    // Obtenez le token
    const accessToken = await getAccessToken();

    const playlistNames = [
      'Pop mix',
      'Rock',
      'Années 80',
      'Années 90',
      'Musique de film',
      'Jazz',
      'R&B',
      'Country',
      'Dubstep',
      'Hip-Hop',
      'Rap',
      'Rap Francais',
      'Rap US',
      'Beurette à chicha',
      'Funk',
      'Années 2000',
      'Afrotrap',
      'Bachata',
      'Latino',
      'Dua Lipa',
    ];

    for (const themeOnly of playlistNames) {
      // Effectuez la recherche de la playlist
      const resultPlaylistId = await searchPlaylistId(accessToken, themeOnly);

      for (const playlistInfo of resultPlaylistId.playlists?.items || []) {
        const playlistId = playlistInfo?.id;
        console.log(playlistId);

        const resultPlaylist = await searchPlaylist(
          accessToken,
          playlistId?.toString() || '',
        );

        // Loop through each track in the playlist
        for (const trackInfo of resultPlaylist.tracks?.items || []) {
          // Get the track information
          const trackId = trackInfo.track?.id;
          const trackName = trackInfo.track?.name;
          const trackArtist = trackInfo.track?.artists?.[0]?.name;
          const previewUrl = trackInfo.track?.preview_url;
          const imageUrl = trackInfo.track?.album?.images?.[0]?.url;

          if (previewUrl) {
            // Utilisez le client Prisma pour insérer dans la base de données
            await prisma.track.create({
              data: {
                track_id: trackId || '',
                name: trackName || '',
                artist: trackArtist || '',
                preview_url: previewUrl || '',
                image_url: imageUrl || '',
                theme: themeOnly || '',
              },
            });

            console.log(`Nom de la piste: ${trackName}`);
            console.log(`Artiste: ${trackArtist}`);
            console.log(`ID Spotify: ${trackId}`);
            console.log(`URL de prévisualisation: ${previewUrl}`);
            console.log(`URL de l'image: ${imageUrl}`);
            console.log('\n');
          }
        }
      }
    }
  } catch (error) {
    console.error("Une erreur s'est produite :", error);
  } finally {
    // Fermez la connexion Prisma à la fin
    await prisma.$disconnect();
  }
};

// Appelez la fonction principale
main();

File: prisma/seed.py
from prisma import Prisma
from access_token import get_access_token
from requests import search_playlist_id, search_playlist
import sqlite3

# Obtenez le token
access_token = get_access_token()

# Initialisez le client Prisma
prisma = Prisma()

playlist_name = ["Pop mix", "Rock", "Années 80", "Années 90", "Musique de film", "Jazz", "R&B", "Country", "Dubstep",
                 "Hip-Hop", "Rap", "Rap Francais", "Rap US", "Beurette à chicha", "Funk", "Années 2000", "Afrotrap",
                 "Bachata", "Latino", "Dua Lipa"]

for theme_only in playlist_name:
    # Effectuez la recherche de la playlist
    result_playlist_id = search_playlist_id(access_token, theme_only)

    for playlist_info in result_playlist_id.get("playlists", {}).get("items", []):
        playlist_id = playlist_info.get("id")
    print(playlist_id)

    result_playlist = search_playlist(access_token, str(playlist_id))

    # Loop through each track in the playlist
    for track_info in result_playlist.get("tracks", {}).get("items", []):
        # Get the track information
        track_id = track_info.get("track", {}).get("id")
        track_name = track_info.get("track", {}).get("name")
        track_artist = track_info.get("track", {}).get("artists", [{}])[0].get("name")
        preview_url = track_info.get("track", {}).get("preview_url")
        image_url = track_info.get("track", {}).get("album", {}).get("images", [{}])[0].get("url")

        if preview_url is not None:
            # Utilisez le client Prisma pour insérer ou mettre à jour la base de données
            prisma.track.upsert(
                where={"track_id": track_id},
                create={"name": track_name, "artist": track_artist, "preview_url": preview_url,
                        "image_url": image_url, "theme": theme_only},
                update={"name": track_name, "artist": track_artist, "preview_url": preview_url,
                        "image_url": image_url, "theme": theme_only},
            )

            print(f"Nom de la piste: {track_name}")
            print(f"Artiste: {track_artist}")
            print(f"ID Spotify: {track_id}")
            print(f"URL de prévisualisation: {preview_url}")
            print(f"URL de l'image: {image_url}")
            print("\n")
